cpp/src/io/parquet/arrow_schema_writer.cpp

```diff
@@ -0,0 +1,362 @@
+  {
+    type_type = flatbuf::Type_Timestamp;
+    // TODO: Verify if this is the correct logic for UTC
+    field_offset =
+      flatbuf::CreateTimestamp(
+        fbb, flatbuf::TimeUnit_NANOSECOND, (utc_timestamps) ? fbb.CreateString("UTC") : 0)
+        .Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::duration_D> or std::is_same_v<T, cudf::duration_s>, void>
+  operator()()
+  {
+    type_type    = flatbuf::Type_Duration;
+    field_offset = flatbuf::CreateDuration(fbb, flatbuf::TimeUnit_SECOND).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::duration_ms>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Duration;
+    field_offset = flatbuf::CreateDuration(fbb, flatbuf::TimeUnit_MILLISECOND).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::duration_us>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Duration;
+    field_offset = flatbuf::CreateDuration(fbb, flatbuf::TimeUnit_MICROSECOND).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<std::is_same_v<T, cudf::duration_ns>, void> operator()()
+  {
+    type_type    = flatbuf::Type_Duration;
+    field_offset = flatbuf::CreateDuration(fbb, flatbuf::TimeUnit_NANOSECOND).Union();
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_fixed_point<T>(), void> operator()()
+  {
+    if constexpr (std::is_same_v<T, numeric::decimal128>) {
+      type_type = flatbuf::Type_Decimal;
+      field_offset =
+        flatbuf::CreateDecimal(fbb, col_meta.get_decimal_precision(), col->type().scale(), 128)
+          .Union();
+    }
+    // cuDF-PQ writer supports ``decimal32`` and ``decimal64`` types, not directly supported by
+    // Arrow without explicit conversion. See more:
+    // https://github.com/rapidsai/cudf/blob/branch-24.08/cpp/src/interop/to_arrow.cu#L155.
+    else {
+      // TODO: Should we fail here or just not write arrow schema?.
+      CUDF_FAIL("Fixed point types smaller than `decimal128` are not supported in arrow schema");
+    }
+  }
+
+  template <typename T>
+  std::enable_if_t<cudf::is_nested<T>(), void> operator()()
+  {
+    // Lists are represented differently in arrow and cuDF.
+    // cuDF representation: List<int>: "col_name" : { "list", "element:int" } (2 children)
+    // arrow schema representation: List<int>: "col_name" : { "list<item:int>" } (1 child)
+    // Hence, we only need to process the second child of the list.
+    if constexpr (std::is_same_v<T, cudf::list_view>) {
+      children.emplace_back(make_arrow_schema_fields(
+        fbb, col->children[1], col_meta.child(1), write_mode, utc_timestamps));
+      type_type    = flatbuf::Type_List;
+      field_offset = flatbuf::CreateList(fbb).Union();
+    }
+
+    // Traverse the struct in DFS manner and process children fields.
+    else if constexpr (std::is_same_v<T, cudf::struct_view>) {
+      std::transform(thrust::make_counting_iterator(0UL),
```

Comment:
This cannot be converted to `zip_iterator` as counter is needed for `col_meta.child(idx)`

Reply:
I think thrust::tabulate would work here, it just requires `children.resize(col->children.size());` beforehand.

Reply:
I think `thrust::tabulate` over `children` would be similar to `for_each` but we need the counter (idx) to access `col_meta.child(..)` unless we make it static in the lambda function which may get messy.